src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
+essentially a one-to-one relationship between `AdtDef` and `DefId`. You can get the `AdtDef` for a
+`DefId` with the [`tcx.adt_def(def_id)`
+query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.adt_def).
+The `AdtDef`s are all interned (as you can see `'tcx` lifetime on it).
+
+We also have an internal map to go from `def_id` to what’s called "Def path". "Def path" is like a
+module path but a bit more rich. For our example, it may be `crate::foo::MyStruct` that identifies
+this definition uniquely. It’s a bit different than a module path because it might include the type
+parameter `T`, which you can't write in normal rust, like `crate::foo::MyStruct::T`. These are used
+in incremental compilation.
+
+### Generics and substitutions
+
+Given a generic type `MyType<A, B, …>`, we may want to swap out the generics `A, B, …` for some
+other types (possibly other generics or concrete types). We do this a lot while doing type
+inference, type checking, and trait solving. Conceptually, during these routines, we may find out
+that one type is equal to another type and want to swap one out for the other and then swap that out
+for another type and so on until we eventually get some concrete types (or an error).
+
+In rustc this is done using the `SubstsRef` that we mentioned above (“substs” = “substitutions”).
+Conceptually, you can think of `SubstsRef` of a list of types that are to be substituted for the
+generic type parameters of the ADT.
+
+`SubstsRef` is a type alias of `List<GenericArg<'tcx>>` (see [`List`
+rustdocs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.List.html)).
+[`GenericArg`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/struct.GenericArg.html)
+is essentially a space-efficient wrapper around
+[`GenericArgKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/enum.GenericArgKind.html),
+which is an enum indicating what kind of generic the type parameter is (type, lifetime, or const).
+Thus, `SubstsRef` is conceptually like a `&'tcx [GenericArgKind<'tcx>]` slice (but it is actually a
+`List`).
+
+So why do we use this `List` type instead of making it really a slice? The reason is that we may
+want to compare subslices of the list of generics for equality. For example, imagine we had a `List`
+`a = [u32, f32]` and  a `List` `b = [f32]`. Now suppose that we have a subslice of `a` : `a2 =
+[f32]`. Recall that lists are interned, so if we want to compare `a` with `b`, we can just compare
+their pointers and see whether they are equivalent or not. But we may *also* want to check the
+equality of `b` and `a2` (i.e. compare subslices of lists). Since the subslice `a2`  would also be
+interned, it would actually end up being the same as `b`, so we would be able to efficiently compare
+them by just comparing the pointers (just like with the full lists). You wouldn’t be able to do that
```

Comment:
I don't understand this paragraph actually. I *think* that it may be a bit misleading, in fact. I believe a `List` is distinct from a slice in two ways:

* it has the `len` "inline", so `&List` is only 32 bits
* as a consequence, it cannot be "subsliced" (that only works if the len is out of line)

This also implies that yes you can check two Lists for equality via `==` (which would be not be possible for ordinary slices). This is precisely because they never represent a "sub-list", only the complete list, which has thus been hashed and interned.

Reply:
Addressed.